src/storage/src/object/disk.rs

```diff
@@ -0,0 +1,431 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::PathBuf;
+
+use bytes::Bytes;
+use tokio::io::{AsyncReadExt, AsyncSeekExt, AsyncWriteExt};
+use tokio::sync::mpsc::UnboundedSender;
+
+use crate::object::{BlockLocation, ObjectError, ObjectMetadata, ObjectResult, ObjectStore};
+
+pub(super) mod utils {
+    use std::fs::Metadata;
+    use std::path::Path;
+
+    use tokio::fs::{create_dir_all, File, OpenOptions};
+
+    use crate::object::{ObjectError, ObjectResult};
+
+    pub async fn ensure_file_dir_exists(path: &Path) -> ObjectResult<()> {
+        if let Some(dir) = path.parent() {
+            // `create_dir_all` will not error even if the directory already exists.
+            create_dir_all(dir).await.map_err(|e| {
+                ObjectError::internal(format!(
```

Comment:
How about adding a new `ObjectError` variant like `Disk` and impl `From<io::Error>` for it?